File: TinyGP.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>

#define MAX_LEN 10000
#define POPSIZE 100000
#define DEPTH 5
#define GENERATIONS 100
#define TSIZE 2
#define PMUT_PER_NODE 0.05
#define CROSSOVER_PROB 0.9

#define ADD 110
#define SUB 111
#define MUL 112
#define DIV 113
#define FSET_START ADD
#define FSET_END DIV

double fitness[POPSIZE];
char pop[POPSIZE][MAX_LEN];
static double x[FSET_START];
static double minrandom, maxrandom;
static char program[MAX_LEN];
static int PC;
static int varnumber, fitnesscases, randomnumber;
static double fbestpop = 0.0, favgpop = 0.0;
static long seed;
static double avg_len;

#define MAX_TARGETS 100  // Maximum fitness cases
double targets[MAX_TARGETS][10];

double run() { /* Interpreter */
    char primitive = program[PC++];
    if (primitive < FSET_START)
        return x[primitive];
    switch (primitive) {
        case ADD: return run() + run();
        case SUB: return run() - run();
        case MUL: return run() * run();
        case DIV: {
            double num = run(), den = run();
            if (fabs(den) <= 0.001)
                return num;
            else
                return num / den;
        }
    }
    return 0.0;
}

int traverse(char buffer[], int buffercount) {
    if (buffer[buffercount] < FSET_START)
        return buffercount + 1;

    switch (buffer[buffercount]) {
        case ADD:
        case SUB:
        case MUL:
        case DIV:
            return traverse(buffer, traverse(buffer, buffercount + 1));
    }
    return 0;
}

void setup_fitness(char *fname) {
    FILE *fp = fopen(fname, "r");
    if (fp == NULL) {
        printf("ERROR: Please provide a data file\n");
        exit(0);
    }

    fscanf(fp, "%d %d %lf %lf %d", &varnumber, &randomnumber, &minrandom, &maxrandom, &fitnesscases);

    if (varnumber + randomnumber >= FSET_START)
        printf("too many variables and constants\n");

    for (int i = 0; i < fitnesscases; i++) {
        for (int j = 0; j <= varnumber; j++) {
            fscanf(fp, "%lf", &targets[i][j]);
        }
    }

    fclose(fp);
}

double fitness_function(char Prog[], double targets[][10]) {
    int i = 0, len;
    double result, fit = 0.0;
    len = traverse(Prog, 0);

    for (i = 0; i < fitnesscases; i++) {
        for (int j = 0; j < varnumber; j++)
            x[j] = targets[i][j];

        strcpy(program, Prog);  // Copy the content of Prog into program
        PC = 0;
        result = run();
        fit += fabs(result - targets[i][varnumber]);
    }
    return -fit;
}

int grow(char buffer[], int pos, int max, int depth) {
    char prim = rand() % 2;
    int one_child;

    if (pos >= max)
        return -1;

    if (pos == 0)
        prim = 1;

    if (prim == 0 || depth == 0) {
        prim = rand() % (varnumber + randomnumber);
        buffer[pos] = prim;
        return pos + 1;
    } else {
        prim = rand() % (FSET_END - FSET_START + 1) + FSET_START;

        switch (prim) {
            case ADD:
            case SUB:
            case MUL:
            case DIV:
                buffer[pos] = prim;
                one_child = grow(buffer, pos + 1, max, depth - 1);
                if (one_child < 0)
                    return -1;
                return grow(buffer, one_child, max, depth - 1);
        }
    }
    return 0;
}

int print_indiv(char buffer[], int buffercounter) {
    int a1 = 0, a2;
    if (buffer[buffercounter] < FSET_START) {
        if (buffer[buffercounter] < varnumber)
            printf("X%d ", buffer[buffercounter] + 1);
        else
            printf("%lf", x[buffer[buffercounter]]);
        return buffercounter + 1;
    }

    switch (buffer[buffercounter]) {
        case ADD:
            printf("(");
            a1 = print_indiv(buffer, buffercounter + 1);
            printf(" + ");
            break;
        case SUB:
            printf("(");
            a1 = print_indiv(buffer, buffercounter + 1);
            printf(" - ");
            break;
        case MUL:
            printf("(");
            a1 = print_indiv(buffer, buffercounter + 1);
            printf(" * ");
            break;
        case DIV:
            printf("(");
            a1 = print_indiv(buffer, buffercounter + 1);
            printf(" / ");
            break;
    }
    a2 = print_indiv(buffer, a1);
    printf(")");
    return a2;
}

char buffer[MAX_LEN];

char *create_random_indiv(int depth) {
    char *ind;
    int len;
    len = grow(buffer, 0, MAX_LEN, depth);

    while (len < 0)
        len = grow(buffer, 0, MAX_LEN, depth);

    ind = (char *)malloc(len);
    memcpy(ind, buffer, len);
    return ind;
}

void create_random_pop(int n, int depth, double fitness[]) {
    for (int i = 0; i < n; i++) {
        char *ind = create_random_indiv(depth);
        strcpy(pop[i], ind);
        fitness[i] = fitness_function(ind, targets);
        free(ind);
    }
}

void stats(double fitness[], int gen) {
    int i, best = rand() % POPSIZE;
    int node_count = 0;
    fbestpop = fitness[best];
    favgpop = 0.0;

    for (i = 0; i < POPSIZE; i++) {
        node_count += traverse(pop[i], 0);
        favgpop += fitness[i];
        if (fitness[i] > fbestpop) {
            best = i;
            fbestpop = fitness[i];
        }
    }
    avg_len = (double)node_count / POPSIZE;
    favgpop /= POPSIZE;
    printf("Generation=%d Avg Fitness=%lf Best Fitness=%lf Avg Size=%lf\nBest Individual: ",
           gen, -favgpop, -fbestpop, avg_len);
    print_indiv(pop[best], 0);
    printf("\n");
    fflush(stdout);
}

int tournament(double fitness[], int tsize) {
    int best = rand() % POPSIZE;
    int competitor;
    double fbest = -1.0e34;

    for (int i = 0; i < tsize; i++) {
        competitor = rand() % POPSIZE;
        if (fitness[competitor] > fbest) {
            fbest = fitness[competitor];
            best = competitor;
        }
    }
    return best;
}

int negative_tournament(double fitness[], int tsize) {
    int worst = rand() % POPSIZE;
    int competitor;
    double fworst = 1e34;

    for (int i = 0; i < tsize; i++) {
        competitor = rand() % POPSIZE;
        if (fitness[competitor] < fworst) {
            fworst = fitness[competitor];
            worst = competitor;
        }
    }
    return worst;
}

char *crossover(char *parent1, char *parent2) {
    int xo1start, xo1end, xo2start, xo2end;
    char *offspring;
    int len1 = traverse(parent1, 0);
    int len2 = traverse(parent2, 0);
    int lenoff;

    xo1start = rand() % len1;
    xo1end = traverse(parent1, xo1start);

    xo2start = rand() % len2;
    xo2end = traverse(parent2, xo2start);

    lenoff = xo1start + (xo2end - xo2start) + (len1 - xo1end);

    offspring = (char *)malloc(lenoff);
    memcpy(offspring, parent1, xo1start);
    memcpy(offspring + xo1start, parent2 + xo2start, (xo2end - xo2start));
    memcpy(offspring + xo1start + (xo2end - xo2start), parent1 + xo1end, (len1 - xo1end));

    return offspring;
}

char *mutation(char *parent, double pmut) {
    int len = traverse(parent, 0);
    int i;
    int mutsite;
    char *parentcopy = (char *)malloc(len);
    memcpy(parentcopy, parent, len);

    for (i = 0; i < len; i++) {
        if ((double)rand() / RAND_MAX < pmut) {
            mutsite = i;
            if (parentcopy[mutsite] < FSET_START)
                parentcopy[mutsite] = rand() % (varnumber + randomnumber);
            else
                switch (parentcopy[mutsite]) {
                    case ADD:
                    case SUB:
                    case MUL:
                    case DIV:
                        parentcopy[mutsite] = rand() % (FSET_END - FSET_START + 1) + FSET_START;
                }
        }
    }
    return parentcopy;
}

void print_parms() {
    printf("-- TINY GP (C version) --\n");
    printf("SEED=%ld\nMAX_LEN=%d\nPOPSIZE=%d\nDEPTH=%d\nCROSSOVER_PROB=%lf\nPMUT_PER_NODE=%lf\nMIN_RANDOM=%lf\nMAX_RANDOM=%lf\nGENERATIONS=%d\nTSIZE=%d\n----------------------------------\n",
           seed, MAX_LEN, POPSIZE, DEPTH, CROSSOVER_PROB, PMUT_PER_NODE, minrandom, maxrandom, GENERATIONS, TSIZE);
}

int main(int argc, char *argv[]) {
    char *fname = "xmxp2.dat"; //zad1 fun1 dzi1 -10 10
    long s = -1;

    if (argc == 3) {
        s = atoi(argv[1]);
        fname = argv[2];
    } else if (argc == 2) {
        fname = argv[1];
    }

    srand(time(NULL));

    setup_fitness(fname);
    for (int i = 0; i < FSET_START; i++)
        x[i] = (maxrandom - minrandom) * ((double)rand() / RAND_MAX) + minrandom;
    create_random_pop(POPSIZE, DEPTH, fitness);

    print_parms();
    stats(fitness, 0);

    for (int gen = 1; gen < GENERATIONS; gen++) {
        if (fbestpop > -1e-5) {
            printf("PROBLEM SOLVED\n");
            exit(0);
        }

        for (int indivs = 0; indivs < POPSIZE; indivs++) {
            if ((double)rand() / RAND_MAX < CROSSOVER_PROB) {
                int parent1 = tournament(fitness, TSIZE);
                int parent2 = tournament(fitness, TSIZE);
                char *newind = crossover(pop[parent1], pop[parent2]);
                double newfit = fitness_function(newind, targets);
                int offspring = negative_tournament(fitness, TSIZE);
                free(pop[offspring]);
                strcpy(pop[offspring], newind);
                fitness[offspring] = newfit;
                free(newind);
            } else {
                int parent = tournament(fitness, TSIZE);
                char *newind = mutation(pop[parent], PMUT_PER_NODE);
                double newfit = fitness_function(newind, targets);
                int offspring = negative_tournament(fitness, TSIZE);
                free(pop[offspring]);
                strcpy(pop[offspring], newind);
                fitness[offspring] = newfit;
                free(newind);
            }
        }
        stats(fitness, gen);
    }

    printf("PROBLEM *NOT* SOLVED\n");
    exit(1);
}

File: lab2 - trigonometrics and calculations/TinyGP.java
/*
 * Original Author:    Riccardo Poli (email: rpoli@essex.ac.uk)
 *
 * Refactored by Jakub Banach and Karol Błaszczak
 * 
 * For the Genetic Programming laboratory classes in the Computer Science and Intelligent Systems course at AGH University of Science and Technology.
 * 
 * Final version for laboratory no.2
 */

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Random;
import java.util.StringTokenizer;

public class TinyGP {
    double[] fitness;
    char[][] pop;
    static Random random = new Random();
    static final int
            ADD = 110,
            SUB = 111,
            MUL = 112,
            DIV = 113,
            SIN = 114,
            COS = 115,
            FSET_START = ADD,
            FSET_END = COS;
    static double[] x = new double[FSET_START];
    static double minRandom, maxRandom;
    static char[] program;
    static int PC;
    static int varNumber, fitnessCases, randomNumber;
    static double fBestPop = 0.0, fAvgPop = 0.0;
    static long seed;
    static double avgLen;
    static final int
            MAX_LEN = 10000,
            POPSIZE = 100000,
            DEPTH = 5,
            GENERATIONS = 500,
            TSIZE = 2;
    public static final double
            PMUT_PER_NODE = 0.05,
            CROSSOVER_PROB = 0.9;
    static double[][] targets;
    static char[] buffer = new char[MAX_LEN];
    static String solution = "";

    public static boolean solved = false;
    long longestProgram = 0L;
    public static String finalSolution;
    public static double finalBestFit = -Double.MAX_VALUE;

    double run() {
        char primitive = program[PC++];
        if (primitive < FSET_START) {
            return x[primitive];
        }
        switch (primitive) {
            case ADD:
                return run() + run();
            case SUB:
                return run() - run();
            case MUL:
                return run() * run();
            case DIV: {
                double num = run(), den = run();
                if (Math.abs(den) <= 0.001)
                    return num;
                else
                    return num / den;
            }
            case SIN:
                return Math.sin(run());
            case COS:
                return Math.cos(run());
        }
        return 0.0; // should never get here
    }

    int traverse(char[] buffer, int bufferCount) {
        return buffer[bufferCount] < FSET_START
                ? ++bufferCount
                : switch (buffer[bufferCount]) {
            case ADD, SUB, MUL, DIV -> traverse(buffer, traverse(buffer, ++bufferCount));
            case SIN, COS -> traverse(buffer, ++bufferCount);
            default -> 0; // should never get here
        };
    }

    void setupFitness(String fileName) {
        try (BufferedReader in = new BufferedReader(new FileReader(fileName))) {
            String line = in.readLine();
            StringTokenizer tokens = new StringTokenizer(line);
            varNumber = Integer.parseInt(tokens.nextToken().trim());
            randomNumber = Integer.parseInt(tokens.nextToken().trim());
            minRandom = Double.parseDouble(tokens.nextToken().trim());
            maxRandom = Double.parseDouble(tokens.nextToken().trim());
            fitnessCases = Integer.parseInt(tokens.nextToken().trim());
            targets = new double[fitnessCases][varNumber + 1];
            if (varNumber + randomNumber >= FSET_START) {
                System.out.println("too many variables and constants");
            }

            for (int i = 0; i < fitnessCases; i++) {
                line = in.readLine();
                tokens = new StringTokenizer(line);
                for (int j = 0; j <= varNumber; j++) {
                    targets[i][j] = Double.parseDouble(tokens.nextToken().trim());
                }
            }
        } catch (FileNotFoundException e) {
            System.out.println("ERROR: Please provide a data file");
            System.exit(0);
        } catch (Exception e) {
            System.out.println("ERROR: Incorrect data format: " + e.getMessage());
            System.exit(0);
        }
    }

    double fitnessFunction(char[] Prog) {
        double result, fit = 0.0;

        traverse(Prog, 0);
        for (int i = 0; i < fitnessCases; i++) {
            if (varNumber >= 0) {
                System.arraycopy(targets[i], 0, x, 0, varNumber);
            }
            program = Prog;
            PC = 0;
            result = run();
            fit += Math.abs(result - targets[i][varNumber]);
        }
        return (-1) * fit;
    }

    int grow(int pos, int max, int depth) {
        if (pos >= max) {
            return -1;
        }
        char prim = pos == 0 ? 1 : (char) random.nextInt(2);

        if (prim == 0 || depth == 0) {
            prim = (char) random.nextInt(varNumber + randomNumber);
            buffer[pos] = prim;
            return pos + 1;
        } else {
            prim = (char) (random.nextInt(FSET_END - FSET_START + 1) + FSET_START);
            buffer[pos] = prim;
            int oneChild = grow(pos + 1, max, depth - 1);
            if (oneChild < 0) {
                return -1;
            }
            switch (prim) {
                case ADD, SUB, MUL, DIV -> {
                    return grow(oneChild, max, depth - 1);
                }
                case SIN, COS -> {
                    return oneChild;
                }
            }
            return 0; // should never get here
        }
    }


    char[] createRandomIndiv() {
        int len;
        do {
            len = grow(0, MAX_LEN, DEPTH);
        } while (len < 0);
        char[] ind = new char[len];
        System.arraycopy(buffer, 0, ind, 0, len);
        return ind;
    }

    char[][] createRandomPop() {
        char[][] pop = new char[POPSIZE][];
        for (int i = 0; i < POPSIZE; i++) {
            pop[i] = createRandomIndiv();
            fitness[i] = fitnessFunction(pop[i]);
        }
        return pop;
    }


    void stats(int gen) {
        int best = random.nextInt(POPSIZE);
        int nodeCount = 0;
        solution = "";
        fBestPop = fitness[best];
        fAvgPop = 0.0;

        for (int i = 0; i < POPSIZE; i++) {
            longestProgram = Long.MIN_VALUE;
            nodeCount += traverse(pop[i], 0);
            fAvgPop += fitness[i];
            StringBuilder stringBuilder = new StringBuilder();
            programOutput(stringBuilder, pop[i], 0);
            String output = stringBuilder.toString();
            long programLength = output.length();
            if (fitness[i] > fBestPop) {
                solution = output;
                fBestPop = fitness[i];
            }
            if (programLength > longestProgram) {
                longestProgram = programLength;
            }
        }
        avgLen = (double) nodeCount / POPSIZE;
        fAvgPop /= POPSIZE;
        System.out.print("Generation=" + gen + " Avg Fitness=" + (-fAvgPop) +
                " Best Fitness=" + (-fBestPop) + " Avg Size=" + avgLen +
                "\nBest Individual: ");
        System.out.println(solution);
        System.out.print("\n");
        System.out.flush();
        if (!solution.equals("") && (-fBestPop) < (-finalBestFit)) {
            finalSolution = solution;
            finalBestFit = fBestPop;
        }
    }

    int tournament() {
        int best = random.nextInt(POPSIZE), competitor;
        double fbest = -1.0e34;

        for (int i = 0; i < TSIZE; i++) {
            competitor = random.nextInt(POPSIZE);
            StringBuilder stringBuilder = new StringBuilder();
            programOutput(stringBuilder, pop[competitor], 0);
            int competitorLength = stringBuilder.toString().length();
            double value = fitness[competitor];
            if (value > fbest && competitorLength < longestProgram) {
                fbest = value;
                best = competitor;
            }
        }
        return best;
    }

    int negativeTournament() {
        int worst = random.nextInt(POPSIZE), competitor;
        double fworst = 1e34;

        for (int i = 0; i < TSIZE; i++) {
            competitor = random.nextInt(POPSIZE);
            double value = fitness[competitor];
            if (value < fworst) {
                fworst = value;
                worst = competitor;
            }
        }
        return worst;
    }

    char[] crossover(char[] parent1, char[] parent2) {
        int len1 = traverse(parent1, 0);
        int len2 = traverse(parent2, 0);

        int xo1start = random.nextInt(len1);
        int xo1end = traverse(parent1, xo1start);

        int xo2start = random.nextInt(len2);
        int xo2end = traverse(parent2, xo2start);

        int lenOff = xo1start + (xo2end - xo2start) + (len1 - xo1end);

        char[] offspring = new char[lenOff];

        System.arraycopy(parent1, 0, offspring, 0, xo1start);
        System.arraycopy(parent2, xo2start, offspring, xo1start, (xo2end - xo2start));
        System.arraycopy(parent1, xo1end, offspring, xo1start + (xo2end - xo2start), (len1 - xo1end));

        return offspring;
    }

    char[] mutation(char[] parent) {
        int len = traverse(parent, 0);
        char[] parentCopy = new char[len];

        System.arraycopy(parent, 0, parentCopy, 0, len);
        for (int i = 0; i < len; i++) {
            if (random.nextDouble() < PMUT_PER_NODE) {
                if (parentCopy[i] < FSET_START) {
                    parentCopy[i] = (char) random.nextInt(varNumber + randomNumber);
                } else
                    switch (parentCopy[i]) {
                        case ADD, SUB, MUL, DIV -> parentCopy[i] =
                                (char) (random.nextInt(DIV - FSET_START + 1) + FSET_START);
                        case SIN, COS -> parentCopy[i] =
                                (char) (random.nextInt(FSET_END - SIN + 1) + SIN);
                    }
            }
        }
        return parentCopy;
    }

    void printParams() {
        System.out.print("-- TINY GP (Java version) --\n");
        System.out.print("SEED=" + seed + "\nMAX_LEN=" + MAX_LEN +
                "\nPOPSIZE=" + POPSIZE + "\nDEPTH=" + DEPTH +
                "\nCROSSOVER_PROB=" + CROSSOVER_PROB +
                "\nPMUT_PER_NODE=" + PMUT_PER_NODE +
                "\nMIN_RANDOM=" + minRandom +
                "\nMAX_RANDOM=" + maxRandom +
                "\nGENERATIONS=" + GENERATIONS +
                "\nTSIZE=" + TSIZE +
                "\n----------------------------------\n");
    }

    public TinyGP(String fileName, long s) {
        fitness = new double[POPSIZE];
        seed = s;
        if (seed >= 0) {
            random.setSeed(seed);
        }
        setupFitness(fileName);
        for (int i = 0; i < FSET_START; i++) {
            x[i] = (maxRandom - minRandom) * random.nextDouble() + minRandom;
        }
        pop = createRandomPop();
    }

    void evolve() {
        solved = false;
        solution = "";
        finalSolution = "";
        finalBestFit = -Double.MAX_VALUE;
        int offspring, parent1, parent2, parent;
        double newFit;
        char[] newInd;
        printParams();
        stats(0);
        for (int gen = 1; gen < GENERATIONS; gen++) {
            if (fBestPop > -1e-5) {
                System.out.print("PROBLEM SOLVED\n");
                solved = true;
                break;
            }
            for (int indivs = 0; indivs < POPSIZE; indivs++) {
                if (random.nextDouble() < CROSSOVER_PROB) {
                    parent1 = tournament();
                    parent2 = tournament();
                    newInd = crossover(pop[parent1], pop[parent2]);
                } else {
                    parent = tournament();
                    newInd = mutation(pop[parent]);
                }
                newFit = fitnessFunction(newInd);
                offspring = negativeTournament();
                pop[offspring] = newInd;
                fitness[offspring] = newFit;
            }
            stats(gen);
        }
        if (!solved) {
            System.out.print("PROBLEM *NOT* SOLVED\n");
        }
    }

    int programOutput(StringBuilder stringBuilder, char[] buffer, int bufferCounter) {
        int a1 = 0;
        if (buffer[bufferCounter] < FSET_START) {
            if (buffer[bufferCounter] < varNumber) {
                stringBuilder.append("X").append(buffer[bufferCounter] + 1);
            } else {
                double value = x[buffer[bufferCounter]];
                String output = value < 0 ? "(" + value + ")" : Double.toString(value);
                stringBuilder.append(output);
            }
            return ++bufferCounter;
        }
        switch (buffer[bufferCounter]) {
            case ADD -> {
                stringBuilder.append("(");
                a1 = programOutput(stringBuilder, buffer, ++bufferCounter);
                stringBuilder.append(" + ");
            }
            case SUB -> {
                stringBuilder.append("(");
                a1 = programOutput(stringBuilder, buffer, ++bufferCounter);
                stringBuilder.append(" - ");
            }
            case MUL -> {
                stringBuilder.append("(");
                a1 = programOutput(stringBuilder, buffer, ++bufferCounter);
                stringBuilder.append(" * ");
            }
            case DIV -> {
                stringBuilder.append("(");
                a1 = programOutput(stringBuilder, buffer, ++bufferCounter);
                stringBuilder.append(" / ");
            }
            case SIN -> {
                stringBuilder.append("sin(");
                a1 = programOutput(stringBuilder, buffer, ++bufferCounter);
                stringBuilder.append(")");
                return a1;
            }
            case COS -> {
                stringBuilder.append("cos(");
                a1 = programOutput(stringBuilder, buffer, ++bufferCounter);
                stringBuilder.append(")");
                return a1;
            }
        }
        int a2 = programOutput(stringBuilder, buffer, a1);
        stringBuilder.append(")");
        return a2;
    }
    public static void main(String[] args) {
        String FileName = "lab2fun2dzi1 -1.570796 1.570796 0.01570796";
        String inputFileName = "./files/2 - function_calculated/" + FileName + ".dat";
        String outputFileName = "./files/3 - tinyGP_calculated/done " + FileName + ".dat";
        long seed = 123456;

        TinyGP tinyGP = new TinyGP(inputFileName, seed);
        tinyGP.evolve();

        if (TinyGP.solved) {
            System.out.println("Writing solution to output file: " + outputFileName);
            writeSolutionToFile(outputFileName, TinyGP.finalSolution);
        } else {
            System.out.println("Writing solution to output file: " + outputFileName);
            writeSolutionToFile(outputFileName, TinyGP.finalSolution);
        }
    }

    private static void writeSolutionToFile(String fileName, String solution) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileName))) {
            writer.write(solution);
        } catch (IOException e) {
            System.err.println("Error writing to the output file: " + e.getMessage());
        }
    }
}